clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7430,6 +7430,271 @@ static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
+  bool HasFilter = ArgsSize && isa<InitListExpr>(Args[0]);
+  if (HasFilter && checkAddIRAttributesFilterListExpr(Args[0], S, CI))
+    return true;
+
+  llvm::SmallVector<PartialDiagnosticAt, 8> Notes;
+  bool HasDependentArg = false;
+  for (unsigned I = HasFilter; I < ArgsSize; I++) {
+    Expr *&E = Args[I];
+
+    if (isa<InitListExpr>(E))
+      return S.Diag(E->getBeginLoc(),
+                    diag::err_sycl_add_ir_attr_filter_list_invalid_arg)
+             << CI;
+
+    if (E->isValueDependent() || E->isTypeDependent()) {
+      HasDependentArg = true;
+      continue;
+    }
+
+    Expr::EvalResult Eval;
+    Eval.Diag = &Notes;
+    if (!E->EvaluateAsConstantExpr(Eval, Context) || !Notes.empty()) {
+      S.Diag(E->getBeginLoc(), diag::err_attribute_argument_n_type)
+          << CI << (I + 1) << AANT_ArgumentConstantExpr;
+      for (auto &Note : Notes)
+        S.Diag(Note.first, Note.second);
+      return true;
+    }
+    assert(Eval.Val.hasValue());
+    E = ConstantExpr::Create(Context, E, Eval.Val);
+  }
+
+  // If there are no dependent expressions, check for expected number of args.
+  if (!HasDependentArg && ArgsSize && (ArgsSize - HasFilter) & 1)
+    return S.Diag(CI.getLoc(), diag::err_sycl_add_ir_attribute_must_have_pairs)
+           << CI;
+
+  // If there are no dependent expressions, check argument types.
+  // First half of the arguments are names, the second half are values.
+  unsigned MidArg = (ArgsSize - HasFilter) / 2 + HasFilter;
+  if (!HasDependentArg)
+    for (unsigned I = HasFilter; I < ArgsSize; ++I)
+      if ((I < MidArg && checkAddIRAttributesNameExpr(Args[I], S, CI)) ||
+          (I >= MidArg && checkAddIRAttributesValueExpr(Args[I], S, CI)))
+        return true;
+  return false;
+}
+
+static bool hasDependentExpr(Expr **Exprs, const size_t ExprsSize) {
+  return std::any_of(Exprs, Exprs + ExprsSize, [](const Expr *E) {
+    return E->isValueDependent() || E->isTypeDependent();
+  });
+}
+
+static bool hasSameSYCLAddIRAttributes(
+    const SmallVector<std::pair<std::string, std::string>, 4> &LAttrs,
+    const SmallVector<std::pair<std::string, std::string>, 4> &RAttrs) {
+  std::set<std::pair<std::string, std::string>> LNameValSet{LAttrs.begin(),
+                                                            LAttrs.end()};
+  std::set<std::pair<std::string, std::string>> RNameValSet{RAttrs.begin(),
+                                                            RAttrs.end()};
+  return LNameValSet == RNameValSet;
+}
+
+SYCLAddIRAttributesFunctionAttr *Sema::MergeSYCLAddIRAttributesFunctionAttr(
+    Decl *D, const SYCLAddIRAttributesFunctionAttr &A) {
+  if (const auto *ExistingAttr =
+          D->getAttr<SYCLAddIRAttributesFunctionAttr>()) {
+    // If there are no dependent argument expressions and the filters or the
+    // attributes are different, then fail due to differing duplicates.
+    if (!hasDependentExpr(A.args_begin(), A.args_size()) &&
```

Comment:
Wow this predicate is... something! :-D It's repeated for all three attributes, so I think this is complex enough to warrant extracting into its own helper function.

Reply:
I really wanted to, but the problem is that most of the functions called are members of the attributes and they have no common base with them, so the helper would have to have to get the return of each of the functions. E.g. we would have something like:
```c++
checkSYCLAddIRAttributesMergability(A.args_begin(), A.args_size(),
                                    ExistingAttr->args_begin(), ExistingAttr->args_size(),
                                    A.getAttributeFilter(), ExistingAttr->getAttributeFilter(),
                                    A.getAttributeNameValuePairs(Context),
                                    ExistingAttr->getAttributeNameValuePairs(Context))
```
which isn't all that elegant either. Alternatively, it could be another "additional member" of the attributes. Would that make sense?

Reply:
Can we use duck typing instead? e.g., use a template type in the helper function so that you can pass arbitrary attributes to it, and rely on the fact that if the given attribute is missing a member function because the user is doing things wrong, they'll get compile errors anyway.

Reply:
Good shout! I have added a function `checkSYCLAddIRAttributesMergability` that does this. Is it roughly what you had in mind?

Reply:
Yup, that's what I had in mind!